stan/math/opencl/rev/arena_matrix_cl.hpp

```diff
@@ -5,30 +5,51 @@
 #include <stan/math/rev/core/chainable_alloc.hpp>
 #include <stan/math/opencl/kernel_generator/is_kernel_expression.hpp>
 #include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/matrix_cl_view.hpp>
 #include <utility>
 
 namespace stan {
 namespace math {
+namespace internal {
+template <typename T>
+class arena_matrix_cl_impl : public chainable_alloc, public matrix_cl<T> {
+ public:
+  using Scalar = typename matrix_cl<T>::Scalar;
+  using type = typename matrix_cl<T>::type;
+
+  template <typename... Args>
+  explicit arena_matrix_cl_impl(Args&&... args)
+      : chainable_alloc(), matrix_cl<T>(std::forward<Args>(args)...) {}
+
+  arena_matrix_cl_impl(const arena_matrix_cl_impl&) = default;
+  arena_matrix_cl_impl(arena_matrix_cl_impl&) = default;
+  arena_matrix_cl_impl(arena_matrix_cl_impl&&) = default;
+};
+
+}  // namespace internal
 
 /**
  * A variant of `matrix_cl` that schedules its destructor to be called, so it
  * can be used on the AD stack.
  */
 template <typename T>
-class arena_matrix_cl : public chainable_alloc, public matrix_cl<T> {
+class arena_matrix_cl {
+ private:
+  internal::arena_matrix_cl_impl<T>* impl_;
+
  public:
   using Scalar = typename matrix_cl<T>::Scalar;
   using type = typename matrix_cl<T>::type;
 
   /**
-   * General constructoer forwards arguments to various `matrix_cl` constructors
-   * and schedules the destructor to be called.
+   * General constructor forwards arguments to various `matrix_cl` constructors.
    * @tparam Args argument types
    * @param args arguments
    */
   template <typename... Args>
   explicit arena_matrix_cl(Args&&... args)
-      : chainable_alloc(), matrix_cl<T>(std::forward<Args>(args)...) {}
+      : impl_(new internal::arena_matrix_cl_impl<T>(
+            std::forward<Args>(args)...)) {}
```

Comment:
Should this constructor have a require like `require_any_not_same_t<arena_matrix_cl, Args...>` So that it's not used for the copy/move constructors?

Reply:
Copy and moeve constructors are explicitly defaulted, so this is not an issue.